common/scala/src/main/scala/whisk/core/containerpool/logging/SplunkLogStore.scala

```diff
@@ -0,0 +1,156 @@
+ */
+class SplunkLogStore(
+  actorSystem: ActorSystem,
+  httpFlow: Option[Flow[(HttpRequest, Promise[HttpResponse]), (Try[HttpResponse], Promise[HttpResponse]), Any]] = None,
+  splunkConfig: SplunkLogStoreConfig = loadConfigOrThrow[SplunkLogStoreConfig]("whisk.logstore.splunk"))
+    extends LogDriverLogStore(actorSystem) {
+  implicit val as = actorSystem
+  implicit val ec = as.dispatcher
+  implicit val materializer = ActorMaterializer()
+
+  private val splunkApi = "/services/search/jobs" //see http://docs.splunk.com/Documentation/Splunk/6.6.3/RESTREF/RESTsearch#search.2Fjobs
+
+  val log = actorSystem.log
+  val maxPendingRequests = 500
+
+  val defaultHttpFlow = Http().cachedHostConnectionPoolHttps[Promise[HttpResponse]](
+    host = splunkConfig.host,
+    port = splunkConfig.port,
+    connectionContext =
+      if (splunkConfig.disableSNI)
+        Http().createClientHttpsContext(AkkaSSLConfig().mapSettings(s => s.withLoose(s.loose.withDisableSNI(true))))
+      else Http().defaultClientHttpsContext)
+
+  override def fetchLogs(activation: WhiskActivation): Future[ActivationLogs] = {
+
+    //example curl request:
+    //    curl -u  username:password -k https://splunkhost:port/services/search/jobs -d exec_mode=oneshot -d output_mode=json -d "search=search index=\"someindex\" | spath=activation_id | search activation_id=a930e5ae4ad4455c8f2505d665aad282 |  table log_message" -d "earliest_time=2017-08-29T12:00:00" -d "latest_time=2017-10-29T12:00:00"
+    //example response:
+    //    {"preview":false,"init_offset":0,"messages":[],"fields":[{"name":"log_message"}],"results":[{"log_message":"some log message"}], "highlighted":{}}
+    val search =
+      s"""search index="${splunkConfig.index}"| spath ${splunkConfig.activationIdField}| search ${splunkConfig.activationIdField}=${activation.activationId.toString}| table ${splunkConfig.logMessageField}"""
+
+    val formatter = DateTimeFormatter.ofPattern("YYYY-MM-dd'T'hh:mm:ss").withZone(ZoneId.of("UTC"))
+    val entity = FormData(
+      Map(
+        "exec_mode" -> "oneshot",
+        "search" -> search,
+        "output_mode" -> "json",
+        "earliest_time" -> formatter.format(activation.start),
+        "latest_time" -> formatter
+          .format(activation.end.plusSeconds(5)))).toEntity //add 5s to avoid a timerange of 0 on short-lived activations
+
+    log.debug("sending request")
+    queueRequest(
+      Post(splunkApi)
+        .withEntity(entity)
+        .withHeaders(List(Authorization(BasicHttpCredentials(splunkConfig.username, splunkConfig.password)))))
+      .flatMap(response => {
+        log.debug(s"splunk API response ${response}")
+        Unmarshal(response.entity)
+          .to[JsObject]
+          .map(r => {
+            ActivationLogs(
+              r.fields("results")
+                .convertTo[JsArray]
+                .elements
+                .map(_.asJsObject.fields(splunkConfig.logMessageField).convertTo[String]))
+          })
```

Comment:
I still believe this should a bit more "modeled", as in:

```scala
case class SplunkResponse(results: Vector[JsObject])

...

Unmarshal(response.entity).to[SplunkResponse].map { r =>
  ActivationLogs(r.results.map(_.fields(splunkConfig.logMessageField).convertTo[String])
}
```

Gives you better compile-time guarantees and possibly better messages in an error case. WDYT?

Reply:
I like it!